### DISCUSSION TITLE
Question: Overriding Pre-fetch count to use with high IO tasks

### DISCUSSION THREAD

**Discussion by ShahBinoy** [celery/celery#7157] [Q&A]
Use case Scenario: Process Millions of files from S3

Approach: 
* List S3 Inventory (Task)
** Split small chunks (Task)
** Each chunk, read n(1000s) of files and process it (Long running Task)

My issue is with the `int` math of pre_fetch_count = concurrency * prefetch limit where I have to only one option, if my concurrency increases, so will my prefetch limit.

Is there a way to directly affect the number of Prefetch count after the math ? or is there a workaround ? 

I want to use Gevent/Eventlet for fetching S3 objects but do not want to pre-fetch more than 1 task at the time.

# Checklist

- [ ] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Category%3A+Documentation%22+)
  for similar or identical bug reports.
- [ ] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [ ] I have included all related issues and possible duplicate issues in this issue
      (If there are none, check this box anyway).

#### Related Issues

- None

#### Possible Duplicates

- None

# Description

# Suggestions

[a comment by an automated account is not shown]

**Comment by clokep** [comment]
> Is there a way to directly affect the number of Prefetch count after the math ? or is there a workaround ?

The way I've handled this in the past is to use different queues for the tasks and then different workers (with different prefetch settings configured) for each queue.

**Comment by ShahBinoy** [comment]
> 

That sounds like a good approach @clokep , my only concern is, because the monkey_patch is happening globally, it would be limiting my worker to 1 concurrency and that would limit the number of concurrent boto3 calls that task invokes.

**Comment by clokep** [comment]
> because the monkey_patch is happening globally, it would be limiting my worker to 1 concurrency and that would limit the number of concurrent boto3 calls that task invokes.

I'm not sure what monkey patch you're talking about?

In my case I was running two completely separate processes for the workers, they had different configurations (I think passed on the command line, but it might have been different settings files).

**Comment by ShahBinoy** [comment]
> I'm not sure what monkey patch you're talking about?
> 
> In my case I was running two completely separate processes for the workers, they had different configurations (I think passed on the command line, but it might have been different settings files).

Sorry, i might have been mixing concerns.

First part of problem is using gevent/eventlet to use "green threads". Here we have a potential to use a large pool of green threads to fetch concurrent S3 objects. 

But this concurrency value, which is controlled at command line parameter `--pool gevent -c 50`, it also increases the prefetch count for workers, which I dont want. 

I want to be able to use high gevent pool but reserve only one task at a time.
